keps/sig-storage/0031-20181029-VolumeSubpathEnvExpansion-apichange.md

```diff
@@ -0,0 +1,351 @@
+
+```
+    env:
+    - name: POD_NAME
+      valueFrom:
+        fieldRef:
+          apiVersion: v1
+          fieldPath: metadata.name
+ 
+   ...
+   
+    volumeMounts:
+    - name: workdir1
+      mountPath: /logs
+      subPath: $(POD_NAME)
+```
+
+to:
+
+```
+    volumeMounts:
+    - name: workdir1
+      mountPath: /logs
+      subPathExpr: $(POD_NAME)
+```
+
+This would then introduce the new element to be processed separately from the `subPath`
+
+### User Stories
+
+## Current workarounds - k8s <=1.9.3
+
+This makes use of symbolical linking to the underlying subpath system
+The symbolic link element was removed after 1.9.3
+
+```
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  labels:
+    app: podtest
+  name: podtest
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: podtest
+  template:
+    metadata:
+      labels:
+        app: podtest
+    spec:
+      containers:
+      - env:
+        - name: POD_NAME
+          valueFrom:
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.name
+        - name: POD_NAMESPACE
+          valueFrom:
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.namespace
+        image: <image>
+        name: podtest
+        volumeMounts:
+        - mountPath: /logs
+          name: workdir
+          subPath: logs
+      initContainers:
+      - command:
+        - /bin/sh
+        - -xc
+        - |
+          LOGDIR=/logs/${POD_NAMESPACE}/${POD_NAME}; mkdir -p ${LOGDIR} && ln -sfv ${LOGDIR} /workdir/logs && chmod -R ugo+wr ${LOGDIR}
+        env:
+        - name: POD_NAME
+          valueFrom:
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.name
+        - name: POD_NAMESPACE
+          valueFrom:
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.namespace
+        image: alpine:3.5
+        name: prep-logs
+        volumeMounts:
+        - mountPath: /logs
+          name: logs
+        - mountPath: /workdir
+          name: workdir
+      volumes:
+      - emptyDir: {}
+        name: workdir
+      - hostPath:
+          path: /logs
+          type: ""
+        name: logs
+```
+
+## Workarounds - k8s >1.9.3
+
+This makes use of relative symbolic linking to the underlying operating system
+
+```
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  name: private-image-test-1
+  labels:
+    application: podtest
+spec:
+  replicas: 2
+  template:
+    metadata:
+      labels:
+        application: podtest
+      name: private-image-test-1
+    spec:
+      initContainers:
+      - command:
+        - /bin/sh
+        - -xc
+        - |
+          LOGDIR=/logs/${POD_NAMESPACE}/${POD_NAME}; mkdir -p ${LOGDIR} && mkdir /workdir/logs && ln -rsfv /var/lib/kubelet/pods/${POD_UID}/volumes/kubernetes.io~empty-dir/workdir/logs ${LOGDIR} && chmod -R ugo+wrx /pods
```

Comment:
To be quite fair I havent actually tried to implement this in any real clusters yet so havent been able to really assess its impact. What it shows is that there is no straight forward workaround after 1.9.3 and is riddled with dangerous behaviour. We are actually going to bypass it and go from 1.9.3 to 1.11 and switch the feature gate on instead.

Reply:
Agree, I have tried to find a workaround with current code and haven't been able to come up with anything that can work.  I think it would be better to leave out this example altogether if it doesn't actually work.